### ISSUE TITLE
Error: spawn [...] ENOENT 

### ISSUE THREAD

**Issue by underyx** [microsoft/vscode-mypy#85]
Ever since installing the extension I'm getting this error, and reselecting the interpreter does not fix it:

```
2023-07-11 09:01:17.355 [info] No interpreter found from setting mypy-type-checker.interpreter
2023-07-11 09:01:17.355 [info] Getting interpreter from ms-python.python extension for workspace /Users/underyx/r2c/app
2023-07-11 09:01:17.357 [info] Interpreter from ms-python.python extension for /Users/underyx/r2c/app: /Users/underyx/r2c/app/server/.venv/bin/python
2023-07-11 09:01:17.358 [info] Using cwd from `python.linting.cwd`.
2023-07-11 09:01:17.361 [info] Server run command: /Users/underyx/r2c/app/server/.venv/bin/python /Users/underyx/.vscode/extensions/ms-python.mypy-type-checker-2023.1.11911749/bundled/tool/lsp_server.py
2023-07-11 09:01:17.361 [info] Server: Start requested.
2023-07-11 09:01:17.449 [info] [Error - 9:01:17 AM] Mypy Type Checker client: couldn't create connection to server.
2023-07-11 09:01:17.449 [info] Launching server using command /Users/underyx/r2c/app/server/.venv/bin/python failed. Error: spawn /Users/underyx/r2c/app/server/.venv/bin/python ENOENT
2023-07-11 09:01:17.449 [error] Server: Start failed: Launching server using command /Users/underyx/r2c/app/server/.venv/bin/python failed. Error: spawn /Users/underyx/r2c/app/server/.venv/bin/python ENOENT
```

even though

```terminal
(semgrep-app-py3.11) underyx@san app % /Users/underyx/r2c/app/server/.venv/bin/python --version
Python 3.11.0
```

This is using v2023.1.11921011 of ms-python.mypy-type-checker.

**Comment by karthiknadig** [comment]
Can you try the following things?
1. Check if this file exists: `/Users/underyx/.vscode/extensions/ms-python.mypy-type-checker-2023.1.11911749/bundled/tool/lsp_server.py`
2. Check if this has the right permissions: `/Users/underyx/r2c/app/server/.venv/bin/python`
3. Check if the linked python has the right permissions: 
```
> readlink /Users/underyx/r2c/app/server/.venv/bin/python
> ls -@l <path-from-symlink>
```
4. Try activating the environment from a terminal and then launching code from there. Sometimes, depending on the where you installed python from, might require running it from an activated terminal. This will allow VS Code itself to have the right PATH to load it.
```
> cd /Users/underyx/r2c/app
> source ./server/.venv/bin/activate
> code .
```

**Comment by underyx** [comment]
Thanks for the detailed instructions!

1. The path you provided does not exist but I think the version number in that one might just have been a mistake, cause this file does exist: `/Users/underyx/.vscode/extensions/ms-python.mypy-type-checker-2023.1.11921011/bundled/tool/lsp_server.py`
2. The permissions on the symlink are `lrwxr-xr-x  1 underyx  staff` which seems right to me
3. The permissions on the symlink target are `-rwxr-xr-x@ 1 underyx  staff` which also seems right
4. launching `code` from within an activated venv did not help either, I get the same error :(

**Comment by underyx** [comment]
Assorted notes, although I don't see how any of these would help, maybe you'll see something useful:

- I'm on macOS 13.4
- I installed python with `asdf`, this project is using python 3.11
- The `.venv` directory is managed by Poetry

**Comment by karthiknadig** [comment]
@underyx Can we try if the server launches at all?
```
/Users/underyx/r2c/app/server/.venv/bin/python /Users/underyx/.vscode/extensions/ms-python.mypy-type-checker-2023.1.11921011/bundled/tool/lsp_server.py
```
This should just run the server and remain blank. It does not output anything it should just wait. If it crashes then something is wrong.

Can you also share the logs from Black extension? up to the `server start requested`.

**Comment by underyx** [comment]
When I run

```
/Users/underyx/r2c/app/server/.venv/bin/python /Users/underyx/.vscode/extensions/ms-python.mypy-type-checker-2023.1.11921011/bundled/tool/lsp_server.py
```

in the vscode integrated terminal, the server works correctly and does not crash. Interrupting it shows it was running this line, which seems correct:

```
Traceback (most recent call last):
  File "/Users/underyx/.vscode/extensions/ms-python.mypy-type-checker-2023.1.11921011/bundled/tool/lsp_server.py", line 641, in <module>
    LSP_SERVER.start_io()
```

I've uploaded full logs from
- Black: https://gist.github.com/underyx/da385c16811f245606eec276f20acd15
- Ruff: https://gist.github.com/underyx/ac3c26670a7b0c27c71f7b8233e7bb0c
- Mypy: https://gist.github.com/underyx/61917404e1ffaeb0e35e51dc0b343010

**Comment by karthiknadig** [comment]
Does setting `mypy-type-checker.interpreter` setting to some other interpreter help? I am not seeing any difference in the logs. The launch code is exactly the same for both black and mypy.

**Comment by underyx** [comment]
I just set it to

```json
"mypy-type-checker.interpreter": ["/usr/bin/python3"]
```

which is my system Python installation, and I get the same error:

```
2023-07-12 16:14:34.744 [info] [Error - 4:14:34 PM] Mypy Type Checker client: couldn't create connection to server.
2023-07-12 16:14:34.744 [info] Launching server using command /usr/bin/python3 failed. Error: spawn /usr/bin/python3 ENOENT
2023-07-12 16:14:34.744 [error] Server: Start failed: Launching server using command /usr/bin/python3 failed. Error: spawn /usr/bin/python3 ENOENT
```

**Comment by karthiknadig** [comment]
I have created a VSIX that explicitly checks existence of both python and the script file. I am hoping it is some issue with how we are using the path themselves. Please try this build:
https://github.com/microsoft/vscode-mypy/suites/14255756408/artifacts/800776463

**Comment by underyx** [comment]
unfortunately, the logs don't seem to reveal anything new 😅 

```
2023-07-12 16:55:53.349 [info] Name: Mypy Type Checker
2023-07-12 16:55:53.349 [info] Module: mypy
2023-07-12 16:55:53.349 [info] Interpreter from setting mypy-type-checker.interpreter: /usr/bin/python3
2023-07-12 16:55:53.349 [info] Using cwd from `python.linting.cwd`.
2023-07-12 16:55:53.349 [info] Interpreter from setting mypy-type-checker.interpreter: /usr/bin/python3
2023-07-12 16:55:53.349 [info] Using cwd from `python.linting.cwd`.
2023-07-12 16:55:53.349 [info] Server run command: /usr/bin/python3 /Users/underyx/.vscode/extensions/ms-python.mypy-type-checker-2023.1.5536209479/bundled/tool/lsp_server.py
2023-07-12 16:55:53.349 [info] Server executable exists: "/usr/bin/python3"
2023-07-12 16:55:53.349 [info] Server executable exists: "/Users/underyx/.vscode/extensions/ms-python.mypy-type-checker-2023.1.5536209479/bundled/tool/lsp_server.py"
2023-07-12 16:55:53.349 [info] Server: Start requested.
2023-07-12 16:55:53.349 [info] [Error - 4:55:53 PM] Mypy Type Checker client: couldn't create connection to server.
2023-07-12 16:55:53.349 [info] Launching server using command /usr/bin/python3 failed. Error: spawn /usr/bin/python3 ENOENT
2023-07-12 16:55:53.349 [error] Server: Start failed: Launching server using command /usr/bin/python3 failed. Error: spawn /usr/bin/python3 ENOENT
```

**Comment by underyx** [comment]
I know sometimes macOS asks special permission prompts for an app to read from certain directories, such as "Would you like to allow VSCode to read your Documents?" but I don't recall ever denying any such prompts. I wonder if such a system setting could cause ENOENT when an extension accesses `/Users/underyx/.vscode/extensions/` though?

**Comment by karthiknadig** [comment]
@underyx It is a node builtin function: https://nodejs.org/api/child_process.html#child_processspawncommand-args-options

**Comment by karthiknadig** [comment]
Try this build, here we launch this directly using `nodejs` API for `spawn` ourselves, instead of letting the LSP client node package. 
https://github.com/microsoft/vscode-mypy/suites/14255756408/artifacts/800776463

If this fails then there is certainly something wrong outside of the extension's control.

**Comment by underyx** [comment]
Hey again @karthiknadig, sorry I forgot about this notification!

The output is the same. Could you maybe share the code snippet that uses `spawn` so I can debug in a node interpreter? I tried to look in the VSIX but the code is minified.

```
2023-07-20 10:06:47.280 [info] Name: Mypy Type Checker
2023-07-20 10:06:47.280 [info] Module: mypy
2023-07-20 10:06:47.280 [info] Interpreter from setting mypy-type-checker.interpreter: /usr/bin/python3
2023-07-20 10:06:47.280 [info] Using cwd from `python.linting.cwd`.
2023-07-20 10:06:47.280 [info] Interpreter from setting mypy-type-checker.interpreter: /usr/bin/python3
2023-07-20 10:06:47.280 [info] Using cwd from `python.linting.cwd`.
2023-07-20 10:06:47.280 [info] Server run command: /usr/bin/python3 /Users/underyx/.vscode/extensions/ms-python.mypy-type-checker-2023.1.5536209479/bundled/tool/lsp_server.py
2023-07-20 10:06:47.280 [info] Server executable exists: "/usr/bin/python3"
2023-07-20 10:06:47.280 [info] Server executable exists: "/Users/underyx/.vscode/extensions/ms-python.mypy-type-checker-2023.1.5536209479/bundled/tool/lsp_server.py"
2023-07-20 10:06:47.280 [info] Server: Start requested.
2023-07-20 10:06:47.280 [info] [Error - 10:06:46 AM] Mypy Type Checker client: couldn't create connection to server.
2023-07-20 10:06:47.280 [info] Launching server using command /usr/bin/python3 failed. Error: spawn /usr/bin/python3 ENOENT
2023-07-20 10:06:47.280 [error] Server: Start failed: Launching server using command /usr/bin/python3 failed. Error: spawn /usr/bin/python3 ENOENT
```

**Comment by karthiknadig** [comment]
That link I shared was probably already broken, they only last for few days. Here is the latest: https://github.com/microsoft/vscode-mypy/suites/14449644692/artifacts/815764999

This is the PR where I am trying to launch it manually: https://github.com/microsoft/vscode-mypy/pull/87

This is where in code we do a start in this extension, which causes the launch:
https://github.com/microsoft/vscode-mypy/blob/1c6c711ff04175aa5026816e3903ee5e74400dd0/src/common/server.ts#L118

That line internally calls this in (`vscode-languageclient`):
https://github.com/microsoft/vscode-languageserver-node/blob/f2ff7d55464a1f58f978cb6635bd8865f050553c/client/src/node/main.ts#L353

**Comment by karthiknadig** [comment]
That means something is preventing this process staying on. It is able to launch, but something instantly kills it or it fails in some way.

In the build I gave you, I tried to manually launch the server using `child_process.spawn`, and give the stdio streams to VS Code. The process is launching, other wise we would have seen "server start error". But it seems like, by the time VS Code starts to communicate with the process it gets killed.

Can you add the following setting to your user settings: `"mypy-type-checker.trace.server": "verbose"`, hopefully that will give us more clues. I will give you another build with little more info.

**Comment by AlexBarev** [comment]
@karthiknadig @underyx I obverserved same erorrs recently for both microsoft/vscode-flake8 and microsoft/mypy. 

1) I forked this pull-request (https://github.com/microsoft/vscode-mypy/pull/87)
2) created `.vsix`
3) installed it in vscode in docker container with this environment:
**OS**: Ubuntu 22.04.3 LTS
**python**: 3.11.5
**code**: 1.82.2 abd2f3db4bdb28f9e95536dfa84d8479f1eb312d x64
**mypy**: 1.5.1 (compiled: yes)

I've set `"mypy-type-checker.trace.server": "verbose"`. Also I set `"mypy-type-checker.interpreter"` and  `"mypy-type-checker.path"` to point to actual executables just in case to be sure there is no problem during sym links resolving.

Here you can see logs output https://gist.github.com/AlexBarev/906d2fe871671bc9462894544d3b4e17.

Any suggestions what to do next?

**Comment by karthiknadig** [comment]
@AlexBarev This can occur if something is writing to the `stdout` on load of python. Is there a way you can create a minimal repro of this?

**Comment by underyx** [comment]
Hi, sorry I forgot about this issue! Sometime in the last two weeks the extension started working, and now I see output like this on the same project as before:

```
2023-10-11 16:05:41.440 [info] ##########Linting Output - mypy##########

2023-10-11 16:05:41.440 [info] semgrep_app/saas/models/deployment.py:257: error: Incompatible types in assignment (expression has type "datetime", variable has type "Column")  [assignment]
Found 1 error in 1 file (checked 1 source file)
```

I am on v2023.4.0 of ms-python.mypy-type-checker now.

**Comment by Alexander-Wilms** [comment]
I encounter this issue after switching from mypy 1.5.1 installed from the openSUSE repos to version 1.7.1 installed in a venv:

```
alexander@PC:~/Technik/Development/advent_of_code_2023> python3.11 -m venv .
alexander@PC:~/Technik/Development/advent_of_code_2023> ./bin/pip3 install -U mypy
Collecting mypy
  Obtaining dependency information for mypy from https://files.pythonhosted.org/packages/af/9a/ad4b219cf27496653c2312407d6a47593f25f5b53e2d163ade5961cac78c/mypy-1.7.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata
  Using cached mypy-1.7.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (1.8 kB)
Collecting typing-extensions>=4.1.0 (from mypy)
  Obtaining dependency information for typing-extensions>=4.1.0 from https://files.pythonhosted.org/packages/24/21/7d397a4b7934ff4028987914ac1044d3b7d52712f30e2ac7a2ae5bc86dd0/typing_extensions-4.8.0-py3-none-any.whl.metadata
  Using cached typing_extensions-4.8.0-py3-none-any.whl.metadata (3.0 kB)
Collecting mypy-extensions>=1.0.0 (from mypy)
  Using cached mypy_extensions-1.0.0-py3-none-any.whl (4.7 kB)
Using cached mypy-1.7.1-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (12.4 MB)
Using cached typing_extensions-4.8.0-py3-none-any.whl (31 kB)
Installing collected packages: typing-extensions, mypy-extensions, mypy
Successfully installed mypy-1.7.1 mypy-extensions-1.0.0 typing-extensions-4.8.0

[notice] A new release of pip is available: 23.2.1 -> 23.3.1
[notice] To update, run: python3.11 -m pip install --upgrade pip
alexander@PC:~/Technik/Development/advent_of_code_2023> ./python3.11 -m pip install --upgrade pip
bash: ./python3.11: Datei oder Verzeichnis nicht gefunden
alexander@PC:~/Technik/Development/advent_of_code_2023> ./bin/python3.11 -m pip install --upgrade pip
Requirement already satisfied: pip in ./lib64/python3.11/site-packages (23.2.1)
Collecting pip
  Obtaining dependency information for pip from https://files.pythonhosted.org/packages/47/6a/453160888fab7c6a432a6e25f8afe6256d0d9f2cbd25971021da6491d899/pip-23.3.1-py3-none-any.whl.metadata
  Using cached pip-23.3.1-py3-none-any.whl.metadata (3.5 kB)
Using cached pip-23.3.1-py3-none-any.whl (2.1 MB)
Installing collected packages: pip
  Attempting uninstall: pip
    Found existing installation: pip 23.2.1
    Uninstalling pip-23.2.1:
      Successfully uninstalled pip-23.2.1
Successfully installed pip-23.3.1
alexander@PC:~/Technik/Development/advent_of_code_2023> ./bin/pip3 show mypy
Name: mypy
Version: 1.7.1
Summary: Optional static typing for Python
Home-page: https://www.mypy-lang.org/
Author: Jukka Lehtosalo
Author-email: jukka.lehtosalo@iki.fi
License: MIT
Location: /home/alexander/Technik/Development/advent_of_code_2023/lib64/python3.11/site-packages
Requires: mypy-extensions, typing-extensions
Required-by:
```

settings.json:
```json
[...]
"mypy.dmypyExecutable": "${workspaceFolder}/bin/dmypy",
[...]
```

```
Mypy extension activated, version 0.2.3
Registering listener for interpreter changed event
Listener registered
[1] Check folder: /home/alexander/Technik/Development/advent_of_code_2023
Activation complete
[1] Received python path from Python extension: /usr/bin/python3
[1] Running dmypy in folder /home/alexander/Technik/Development/advent_of_code_2023
/home/alexander/Technik/Development/advent_of_code_2023/bin/dmypy --status-file /home/alexander/.var/app/com.visualstudio.code/config/Code/User/workspaceStorage/6aa9a480d6ce8ba18f940fde3c399e28/matangover.mypy/dmypy-ef52036a7175d483d910011887fe64b438e354e4-169.json run --log-file /home/alexander/.var/app/com.visualstudio.code/config/Code/User/workspaceStorage/6aa9a480d6ce8ba18f940fde3c399e28/matangover.mypy/dmypy-ef52036a7175d483d910011887fe64b438e354e4.log -- . --show-column-numbers --no-error-summary --no-pretty --no-color-output --python-executable /usr/bin/python3
[1] Error running mypy in /home/alexander/Technik/Development/advent_of_code_2023: Error: spawn /home/alexander/Technik/Development/advent_of_code_2023/bin/dmypy ENOENT
```
